tests/instruments/hp/test_hp3478a.py

```diff
@@ -0,0 +1,143 @@
+
+VALID_CAL_DATA = [
+        0, 0, 0, 0, 3, 0, 8, 2, 15, 4, 4, 0, 13, 11, 0, 0,
+        0, 0, 3, 3, 2, 15, 5, 3, 0, 14, 0, 0, 0, 0, 0, 0,
+        3, 2, 15, 4, 0, 0, 14, 7, 9, 9, 9, 9, 9, 7, 2, 0,
+        15, 3, 12, 10, 11, 9, 9, 9, 9, 9, 9, 2, 0, 14, 15, 14,
+        9, 12, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 15, 15, 9,
+        9, 8, 6, 0, 9, 2, 14, 14, 0, 12, 10, 12, 9, 9, 9, 9,
+        9, 5, 1, 12, 0, 5, 14, 10, 13, 9, 9, 9, 9, 9, 8, 1,
+        12, 1, 15, 1, 10, 12, 0, 0, 0, 0, 0, 0, 1, 12, 15, 3,
+        0, 14, 0, 9, 9, 9, 9, 9, 9, 1, 12, 13, 2, 14, 9, 15,
+        9, 9, 9, 9, 9, 9, 1, 12, 13, 4, 2, 10, 9, 9, 9, 9,
+        9, 9, 9, 1, 12, 14, 12, 13, 9, 5, 9, 9, 9, 9, 9, 9,
+        1, 12, 2, 1, 15, 10, 10, 0, 0, 0, 0, 4, 2, 3, 0, 0,
+        3, 12, 14, 7, 0, 0, 0, 0, 0, 4, 3, 0, 1, 12, 3, 14,
+        8, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 15, 15, 9, 9,
+        8, 6, 0, 9, 3, 14, 3, 1, 1, 12, 0, 0, 0, 0, 0, 0,
+        0, 0, 0, 0, 0, 0, 15, 15, 0, 0, 0, 0, 0, 0, 0, 0
+    ]
+
+
+def convert_cal_data_to_cal_read_xfers(cal_data):
+    # Convert cal_data into 256 single byte 'W' read operations.
+    cal_read_xfers = []
+    for addr, value in enumerate(cal_data):
+        cal_read_xfers.append([bytes([ord('W'), addr]), bytes([value+64])])
+
+    return cal_read_xfers
+
+
+def convert_cal_data_to_cal_write_xfers(cal_data):
+    # Convert cal_data into 256 single byte 'X' write operations.
+    cal_write_xfers = []
+    for addr, value in enumerate(cal_data):
+        cal_write_xfers.append([bytes([ord('X'), addr, value]), None])
+
+    return cal_write_xfers
+
+# ============================================================
+# TESTS
+# ============================================================
+
+
+def test_calibration_enabled():
+    with expected_protocol(
+            HP3478A,
+            [
+                (b"B", b'\x00\x20\x00\x00\x00'),        # cal_enable bit is set
+                (b"B", b'\x00\x00\x00\x00\x00')
+            ],
+    ) as instr:
+        assert instr.calibration_enabled
+        assert not instr.calibration_enabled
+
+
+def test_calibration_data_getter():
+    cal_read_xfers = convert_cal_data_to_cal_read_xfers(VALID_CAL_DATA)
+
+    with expected_protocol(
+            HP3478A,
+            cal_read_xfers
+    ) as instr:
+        cal_data = instr.calibration_data
+        assert instr.verify_calibration_data(cal_data)
+
+
+def test_calibration_data_setter():
+    valid_cal_write_xfers = convert_cal_data_to_cal_write_xfers(VALID_CAL_DATA)
+
+    with expected_protocol(
+            HP3478A,
+
+            # setter fail due to cal_enable cleared
+            [
+                (b"B", b'\x00\x00\x00\x00\x00'),
+            ] +
+
+            # setter pass
+            [
+                (b"B", b'\x00\x20\x00\x00\x00'),
+            ] + valid_cal_write_xfers +
+
+            # setter fail due to invalid data
+            [
+                (b"B", b'\x00\x20\x00\x00\x00'),
+            ]
```

Comment:
I'd split it in three different tests, that way you know why the setter failed and not only, that it failed. If the first part fails, you do not know, whether the other parts work as expected.
In fact, pytest recommends to do a single assert in a test.

If you want to test several outcomes of some action, you can do the action in a fixture and have several tests, each testing one outcome. You can use classes (starting with "Test") to group tests together.

Reply:
Done.